### ISSUE TITLE
Using dot product as a proxy for probability in NCEAverage

### ISSUE THREAD

**Issue by vinsis** [HobbitLong/CMC#48]
Hi, it seems that you are using the dot product between vectors from two views as a proxy for unknown distribution denoted as p<sub>d</sub> in your paper [here](https://github.com/HobbitLong/CMC/blob/master/NCE/NCEAverage.py#L41). In other words, your h<sub>θ</sub> is the dot product. Theoretically any h<sub>θ</sub> can work so it's all good. 

But doesn't it force the two representations to be similar? I understand the two representations should have high mutual information. But it is not the same as having the two vectors in similar directions.

Obviously it worked out pretty well. But do you think having a parameterized `NCEAverage` loss would have allowed for more representations with not so similar directions but still having high MI?

Thank you again!

**Comment by HobbitLong** [comment]
Hi, @vinsis ,

From the ImageNet experiment, there is a linear projection layer between the representation and contrastive loss. 

You can think you are maximizing mutual information between representations **before** the projection. The projection is similar to a reparameterization and it works in such a way that inner product could play the role.

**Comment by vinsis** [comment]
I am sorry I am not sure I understand what you mean by `a linear projection layer between the representation and contrastive loss`. Do you mean something like the below in [model specification](https://github.com/HobbitLong/CMC/blob/master/models/alexnet.py#L72-L74)?

```python
        self.fc8 = nn.Sequential(
            nn.Linear(4096 // 2, feat_dim)
        )
```

One more thing I noticed is that NCE learns a distribution by learning to classify samples as real or fake. In this case, a _sample_ is dot product between v<sub>1</sub><sup>i</sup> and v<sub>2</sub><sup>j</sup>. In other words the classifier is trying to learn which dot products are real (aka come from the same image) and which ones are fake. We could extend it to any inner product space, not just dot product and __possibly__ get diverse representations while preserving high MI.

**Comment by HobbitLong** [comment]
I originally meant [this one](https://github.com/HobbitLong/CMC/blob/master/models/resnet.py#L126). But yours is also a good example.

Though I don't know which product you want to try, I agree there should be more possibilities.

**Comment by vinsis** [comment]
Thanks again @HobbitLong. Closing it now.
